### ISSUE TITLE
关于sampling_function跑不通

### ISSUE THREAD

**Issue by jiaohuix** [Adaxry/ss_on_decoding_steps#1]
您好，非常感谢您优秀的工作！ 只是我在复制您readme代码的时候遇到了很多错误，我做出修改后，虽然可以运行了。但是我很奇怪您代码中的tril起到了什么作用？tril后thresholds = threshold_table[tgt_lengths]只剩下[max_seq_len-1]个元素，其中前batch_size+1个元素非0，后面全是0，在做mask的时候，batchsize+1后面的位置都是False，也就是全部取了first_decoder_outputs，这与论文里说的沿着解码步golden-truth概率递减的说法冲突了？方便的话请大佬解释下，谢谢！！！

```
import torch


def sampling_function(first_decoder_outputs, first_decoder_inputs, max_seq_len, tgt_lengths):


    # indexs of decoding steps
    t = torch.arange(0, max_seq_len - 1)
    # differenct sampling strategy based on decoding steps
    if sampling_strategy == "exponential":
        threshold_table = exp_radix ** t
    elif sampling_strategy == "sigmoid":
        threshold_table = sigmoid_k / (sigmoid_k + torch.exp(t / sigmoid_k))
    elif sampling_strategy == "linear":
        threshold_table = torch.max(epsilon, 1 - t / max_seq_len)
    else:
        raise  ValueError("Unknown sampling_strategy %s" % sampling_strategy)

    # convert threshold_table to [batch_size, seq_len]
    # threshold_table = threshold_table.unsqueeze_(0).repeat(max_seq_len, 1).tril() # [max_seq_len,max_seq_len-1], 上三角为0
    # thresholds = threshold_table[tgt_lengths] # [max_seq_len-1] ,前tgt_lens+1个元素非0
    # thresholds = thresholds.view(-1, max_seq_len-1) # [1, max_seq_len-1]
    # thresholds = thresholds[:, :seq_len]  # [1, seq_len]

    thresholds = threshold_table.unsqueeze(0).repeat(tgt_lengths,1)[:,:seq_len]

    # conduct sampling based on the above thresholds
    random_select_seed = torch.rand([batch_size, seq_len])
    input_mask = (random_select_seed < thresholds).unsqueeze(-1) # [bsz,seq_len,1]
    second_decoder_inputs = torch.where(input_mask, first_decoder_inputs, first_decoder_outputs)

    return second_decoder_inputs

if __name__ == '__main__':
    batch_size = 4
    seq_len = 30 # current batch
    max_seq_len = 1024 # global max seq len
    hid_size = 20
    sampling_strategy = "exponential"
    # exp_radix = 0.99
    exp_radix = 0.7
    sigmoid_k = 20
    epsilon = 0.2
    first_decoder_outputs = torch.rand(batch_size,seq_len,hid_size)
    first_decoder_inputs = torch.rand(batch_size,seq_len,hid_size)


    res= sampling_function(first_decoder_outputs,first_decoder_inputs,max_seq_len=max_seq_len,tgt_lengths=batch_size)
    print(res)
```

**Comment by jiaohuix** [comment]
已解决，打扰啦。
def sampling_function(
        first_decoder_inputs, # [sos, input_ids] 
        first_decoder_outputs,  # [input_ids, eos]
        max_seq_len, 
        tgt_lengths,
        sampling_strategy = "exponential", # 最高
        epsilon = 0.2,
        exp_radix = 0.99,
        sigmoid_k = 20,
        
        ):
    '''
    conduct scheduled sampling based on the index of decoded tokens 
    param first_decoder_outputs: [batch_size, seq_len, hidden_size], model prediections 
    param first_decoder_inputs: [batch_size, seq_len, hidden_size], ground-truth target tokens
    param max_seq_len: scalar, the max lengh of target sequence
    param tgt_lengths: [batch_size], the lenghs of target sequences in a mini-batch
    '''
    batch_size, seq_len, _ = first_decoder_outputs.shape
    # 将outputs往后挪:  [input_ids, <eos>] -> [<sos>, input_ids]
    first_decoder_outputs = torch.cat((first_decoder_inputs[:, :1, :],  # <sos>
                                       first_decoder_outputs[:, :-1, :]), # input_ids - <eos>
                                       dim=1)

    # indexs of decoding steps
    t = torch.arange(0, max_seq_len) # [start,end)

    # differenct sampling strategy based on decoding steps
    assert sampling_strategy in ["exponential", "sigmoid", "linear"], "Unknown sampling_strategy %s" % sampling_strategy
    if sampling_strategy == "exponential":
        threshold_table = exp_radix ** t  
    elif sampling_strategy == "sigmoid":
        threshold_table = sigmoid_k / (sigmoid_k + torch.exp(t / sigmoid_k ))
    elif sampling_strategy == "linear":        
        threshold_table = torch.max(torch.tensor(epsilon), 1 - t / max_seq_len)


    # convert threshold_table[max_seq_len,max_seq_len] to thresholds [batch_size, seq_len]
    threshold_table = threshold_table.unsqueeze_(0).repeat(max_seq_len, 1).tril() # 下三角
    thresholds = threshold_table[tgt_lengths].view(-1, max_seq_len)
    thresholds = thresholds[:, :seq_len]  # [bsz, seq_len]  # 截断到当前batch的最大序列长度，并且pad部分也不算

    # conduct sampling based on the above thresholds 
    random_select_seed = torch.rand([batch_size, seq_len,1]) #[bsz,seq_len,1] 
    thresholds = thresholds.view([*thresholds.shape,1])
    second_decoder_inputs = torch.where(random_select_seed < thresholds, 
                                        first_decoder_inputs, 
                                        first_decoder_outputs)

    return second_decoder_inputs


if __name__ == "__main__":
    batch_size = 3 
    seq_len = 5  
    hidden_size = 10 
    max_seq_len = 7  

    tgt_lengths = torch.tensor([4, 5, 3])
    first_decoder_inputs = torch.randn(batch_size, seq_len, hidden_size)
    first_decoder_outputs = torch.randn(batch_size, seq_len, hidden_size) 
    print("first_decoder_outputs",first_decoder_outputs.shape)

    second_decoder_inputs = sampling_function(first_decoder_inputs, first_decoder_outputs, max_seq_len, tgt_lengths, sampling_strategy="linear")
    print("second_decoder_inputs",second_decoder_inputs.shape)
